File: 1937-maximum-number-of-points-with-cost/1937-maximum-number-of-points-with-cost.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        def left_sum(arr):
            left = [arr[0]] + [0]*(n-1)
            for j in range(1, n):
                left[j] = max(left[j-1]-1, arr[j])
            return left
        def right_sum(arr):
            right = [0]*(n-1) + [arr[-1]]
            for j in range(n-2, -1, -1):
                right[j] = max(right[j+1]-1, arr[j])
            return right
        
        m, n = len(points), len(points[0])
        prev = points[0]
        for i in range(1, m):
            left_row, right_row, cur = left_sum(prev), right_sum(prev), [0]*n
            for j in range(n):
                cur[j] = points[i][j] + max(left_row[j], right_row[j])
            prev = cur[:]
        return max(prev)

File: 1937-maximum-number-of-points-with-cost/1937-maximum-number-of-points-with-cost.java
class Solution {
    public long maxPoints(int[][] points) {
        int m=points.length;
        int n=points[0].length;
        long[] cur = new long[n];
        long[] prev = new long[n];
        for(int i=0; i<m; i++){
            long maxVal = 0;
            for(int j=0; j<n; j++){
                maxVal = Math.max(maxVal-1, prev[j]);
                cur[j] = maxVal;
            }
            for(int j=n-1; j>=0; j--){
                maxVal = Math.max(maxVal-1, prev[j]);
                cur[j] = Math.max(maxVal, cur[j])+points[i][j];
            }
            long[] temp = prev;
            prev = cur;
            cur = temp;
        }
        return Arrays.stream(prev).max().getAsLong();
    }
}